File: 22FALL/22.generate-parentheses.py
#
# @lc app=leetcode id=22 lang=python3
#
# [22] Generate Parentheses
#

# @lc code=start
class Solution:
    def generateParenthesis(self, n: int):
        ans = []

        def backtrack(tmp, l, r, lrest,rrest):
            if rrest == 0 and lrest == 0 and l == n and r == n:
                ans.append(tmp)
                return
            elif rrest == 0: 
                return
            if lrest > 0:
                backtrack(tmp + '(', l+1,r, lrest - 1, rrest)
            if l > r:
                backtrack(tmp + ')', l, r+1, lrest, rrest-1)
        backtrack('', 0,0, n, n)
        return ans
# @lc code=end

File: VSCode_CPP/22.generate-parentheses.cpp
/*
 * @lc app=leetcode id=22 lang=cpp
 *
 * [22] Generate Parentheses
 */

// @lc code=start
class Solution {
public:
    void bt(string &s, int left, int right, vector<string> &ans) {
        if (left == right && left == 0) {
            ans.push_back(s);
            return;
        }
        if (left > 0) {
            s.push_back('(');
            bt(s, left-1, right, ans);
            s.pop_back();
        }
        if (right > left) {
            s.push_back(')');
            bt(s, left, right-1, ans);
            s.pop_back();
        }
    }
    vector<string> generateParenthesis(int n) {
        string s;
        vector<string> ans;
        bt(s, n, n, ans);
        return ans;
    }
};
// @lc code=end

